aws/acmpca_test.go

```diff
@@ -0,0 +1,103 @@
+package aws
+
+import (
+	"testing"
+	"time"
+
+	awsgo "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/acmpca"
+	"github.com/gruntwork-io/cloud-nuke/util"
+	"github.com/gruntwork-io/go-commons/errors"
+	"github.com/stretchr/testify/assert"
+)
+
+// createTestACMPCA will create am ACMPCA and return its ARN.
+func createTestACMPCA(t *testing.T, session *session.Session, name string) *string {
```

Comment:
Sanity check: Do you have a sense on how much it would cost to run this test? It appears that the cost of ACM Private CA is $400 / month, pro rated to the number of X (days or hours?) that it is active. Does that active period include the days it is waiting to be deleted (the 7 day minimum)? If so, that can rack up quickly for us if we are running cloud nuke tests for each PR.

Depending on how much it costs, can you add an environment variable flag so that we can control when we run the tests related to the ACM PCA functionality (and not on every PR)?

Reply:
I am currently talking to our AWS Enterprise contact as indeed their documentation and support is not clear on this and racked up a significant amount of money just for starting/stopping a CA.

I will add some environment variable and this test will be opt-in in all cases so you have to make a concious decision to run it.

Reply:
We heard back from our contact. The billing is pro-rated and does reconciliation after 24-48 hours and you only pay for those hours.


### Hypothesis from Support-Agent

> Hi Jan
It looks like the billing is fluctuating due to the 24-48 hour delay between the billing console and reality. When a certificate is created, the full amount for the month is added to the bill and this only changes to the pro-rated amount after the certificate has been deleted. I'll keep the internal ticket open with the service team just to confirm that this is exactly what is going on. In terms of your other questions, here are the answers:
>- How does AWS want to do pricing for short-lived CAs? (i.e. can we enable a testing pipeline or not)
->>>> I will forward this request to the service team and I'll update you as soon as I hear back from them.
>- Is there an absolute upper service quota for CAs in an account? (I see 200 quoted right now)
->>>> The quota of 200 is the default, and we can request an increase should you require it, but we would need a detailed use case for this request. 
>- Do deleted CAs (with their 7 - 30 day pending period) count towards the service quota?
->>>> Yes, deleted Private CAs count towards the quota until the end of the restoration period. 


### Answer from AWS PCA Service Team confirming hypothesis from Support-Agent above

Quote

>Thank you for your patience. I've just heard back from the team and my original answer about the delay between the billing console and reality is correct. Because the invoice for each month is issued a few days after the end of the month, by the time the bill is issued, everything will have refreshed and the bill will reflect correctly. For the current month, the "real" bill so far is around $7.

>For some more context on how we charge for short term PCAs, the team provided the following:

>AWS PCA charge customer based on the time period: from the time that CA was created to the time CA was deleted.

>Facts:

>Customer created 22 CAs in FRA region in July. 1 of them are in free trial. 
>Customer create CA, and delete them in a short period of time. --> this is fine, nothing wrong.

>Customer check bill of July, which is not the final statement. There are delays between customer deleted the CA and the >billing team calculate the bill based on latest data.

>If customer check right now, the bill will become about $168. 

>When customer finally get bill in August, the total charge will be around $8. Unless they created new CAs

>Considering this information, the best way to monitor the bill for PCA specifically would be to log the hours and calculate this manually on your side throughout the month.

Reply:
So how do we think about the pricing here?

If we have this test case and it runs, say, nightly for a month, what sort of fee are we talking? What if we get 20 PRs in this repo, and each one does 20 test runs in a month?

I just want to be extra sure we don't end up with a massive AWS bill as a surprise from adding this new functionality!

Reply:
@brikis98 I think we can go with this for now. @weitzjdevk implemented a feature flag to disable the tests by default. We can merge with that disabled.

Separately, I'll run an experiment to enable that test locally once. Then, we can check how much our bill was for August and make our decision on whether to continually test it or not based on that bill cost. Does that seem reasonable?

Reply:
> If we have this test case and it runs, say, nightly for a month, what sort of fee are we talking? What if we get 20 PRs in this repo, and each one does 20 test runs in a month?

tl;dr: Price per run approx `$1.50`

So the rough calculation how this works.

Monthly charge (for CA only): `$400`
Month has hours: `720`

Price (CA) per hour approx. `400/720` approx `$ 0.6`
Issue a certificate charge: `$0.75`

Therefore roughly:

Price per run approx `$1.50`

Timeline (when starting the first day of the month)
1. Day1 (first of the month): `$0`
2. Run it
3. Delete it
3. Day1 bill: `$0`
4. Day2 bill: `$400 + 0.75`
5. Day3 bill: `$400 + 0.75`
6. Day4 bill: `$400/720 + 0.75`

When you do the run in the middle of the month

1. Day15 (middle of the month): `$0`
2. Run it
3. Delete it
3. Day15 bill: `$0`
4. Day16 bill (middle of month. Only 15 more days for pro-rating): `$400 * 15 / 30 + 0.75 = $200 + 0.75`
5. Day17 bill: `$200 + 0.75`
6. Day18 bill: `$400/720 + 0.75`

Reply:
I have confirmed this cost in our account as well. I think 0.60 per run is a tad bit expensive to be continuously running the tests, but selectively running it everytime this feature changes seems reasonable, so the env var based feature switch is satisfactory.